### PR TITLE
[wgpu-hal] gles/wgl: Migrate from ancient/unmaintained `winapi` to `windows-rs`

### PR DESCRIPTION
**Pull request by MarijnS95**
**Connections**
Contributes to #3207
Same as PR #5956 but for GLES/WGL, instead of the Dx12 backend.

**Description**
Replaces ancient, deprecated, unmaintained `winapi` crate with `windows-rs`.  The full justification is in the linked issue and PR above.

**Testing**

```
WGPU_BACKEND=gles cargo run --bin wgpu-examples shadow
```

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [ ] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

File: CHANGELOG.md
Comment (Wumpf) [inline CHANGELOG.md:7]: oh, nice catch 😄

File: wgpu-hal/Cargo.toml
Comment (Wumpf) [inline wgpu-hal/Cargo.toml]: since there's more `windows` crate uses coming, let's put that into the workspace `Cargo.toml` (make sure it's also set with default-features false in there, otherwise it opts in default features for everything with now way to opt out)

```suggestion
windows = { workspace = true, optional = true }
```

**Review by Wumpf — APPROVED** [review]
Great stuff, thank you so much for taking care of the windows-rs migration everywhere!

Imho good to go after the workspace dependency nit (see comment)

File: wgpu-hal/Cargo.toml
Comment (MarijnS95) [reply wgpu-hal/Cargo.toml]: As before, I'm still confused by this unused yet _partially_ maintained workspace dependencies table :upside_down_face: 

Why are some but not all dependencies inherited from the workspace?

Comment (MarijnS95) [reply wgpu-hal/Cargo.toml]: Also, do folks want to start with Windows 0.57 or Windows 0.58? The latter has MSRV 1.70 but it seems `wgpu` is already on `1.76`.

Note that `default-features` only turns off the `std` feature: I'm not exactly sure what that implies but we may get worse error formatting - at least it was like that years ago.

File: Cargo.lock
Comment (Wumpf) [inline Cargo.lock:1525]: might be worth digging what still uses 0.54 here and whether we could avoid the duplication, but doesn't have to be now

File: wgpu-hal/Cargo.toml
Comment (Wumpf) [reply wgpu-hal/Cargo.toml]: >  Why are some but not all dependencies inherited from the workspace?

wgpu is older than the ability to do workspace inheritance of crates and nobody every moved stuff over consistently 🤷 

> Also, do folks want to start with Windows 0.57 or Windows 0.58? The latter has MSRV 1.70 but it seems wgpu is already on 1.76.

Any reason in favor of Windows 0.57? As you pointed out from MSRV pov that's not an issue, so we should rather stay up to date? If there's a reason to stick to the older we also have to make this known to dependabot, otherwise an update will creep in latest next week ;-) 

> Note that default-features only turns off the std feature: I'm not exactly sure what that implies but we may get worse error formatting - at least it was like that years ago.

Good to know, maybe worth a comment on it in case you can confirm? Generally good practice for a library like wgpu not to drag in default features when they're not needed even if it's something like improved error handling, after all a user of wgpu can still do that just fine. But the other way round a user of wgpu wouldn't be able to opt-out of those default features at all

File: Cargo.lock
Comment (cwfitzgerald) [reply Cargo.lock:1525]: Flagging this as a blocker for ff - i think we should be using as wide of a range dependency on windows as we can pull off, as having duplicates can range from being unfortunate to a hard blocker (Firefox)

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
See comment above

File: Cargo.lock
Comment (ErichDonGubler) [reply Cargo.lock:1525]: This is actually easy to answer with `cargo tree`:

```
$ cargo tree -i windows@0.54.0 -i windows@0.58.0 --all-features --all-targets

windows v0.54.0
└── generator v0.8.1
    └── loom v0.7.2
        └── tracy-client v0.17.1
            └── wgpu-benchmark v22.0.0 (E:\wgpu\benches)

windows v0.58.0
├── gpu-allocator v0.27.0
│   └── wgpu-hal v22.0.0 (E:\wgpu\wgpu-hal)
│       ├── wgpu v22.0.0 (E:\wgpu\wgpu)
│       │   ├── wgpu-benchmark v22.0.0 (E:\wgpu\benches)
│       │   ├── wgpu-examples v22.0.0 (E:\wgpu\examples)
│       │   ├── wgpu-info v22.0.0 (E:\wgpu\wgpu-info)
│       │   └── wgpu-test v22.0.0 (E:\wgpu\tests)
│       │       [dev-dependencies]
│       │       └── wgpu-examples v22.0.0 (E:\wgpu\examples)
│       └── wgpu-core v22.0.0 (E:\wgpu\wgpu-core)
│           ├── player v22.0.0 (E:\wgpu\player)
│           └── wgpu v22.0.0 (E:\wgpu\wgpu) (*)
└── wgpu-hal v22.0.0 (E:\wgpu\wgpu-hal) (*)
```

Comment (ErichDonGubler) [reply Cargo.lock:1525]: `windows` 0.52 is what's currently in use in `mozilla-central`, and it has explicit use by _many_ crates. I'm not sure how easy this is to upgrade in `mozilla-central`, but I suspect it's non-trivial, _unless_ `windows` 0.52 is source-compatible _enough_ with 0.58.0.

CC @glandium and @jimblandy, who I'm sure would be interested in the pressure we're feeling on this front.

File: wgpu-hal/Cargo.toml
Comment (MarijnS95) [reply wgpu-hal/Cargo.toml]: > wgpu is older than the ability to do workspace inheritance of crates and nobody every moved stuff over consistently 🤷

It was previously told to me that workspace dependency inheritance was reverted, but the workspace table was kept in place so that it could ever get reenabled.

> Any reason in favor of Windows 0.57? As you pointed out from MSRV pov that's not an issue, so we should rather stay up to date? If there's a reason to stick to the older we also have to make this known to dependabot, otherwise an update will creep in latest next week ;-)

None really; except that I started the conversion on `0.57`. A lot of _my_ high-level API improvements in `Dxgi` via metadata annotations finally landed in `windows 0.58`, too. All dependencies like `gpu-allocator` have released with `windows 0.58` in their range as well.

---

I'll double-check the purpose of the `std` feature before marking this thread as resolved.

File: Cargo.lock
Comment (MarijnS95) [reply Cargo.lock:1525]: > might be worth digging what still uses 0.54

No need to dig, it's pretty easy to see from the `Cargo.lock` context (or `cargo tree` otherwise) that this is [`generator`](https://crates.io/crates/generator).

`cargo tree` however points out that it's a seemingly-irrelevant dependency, exclusively involved in the `wgpu-benchmark` tree and not clobbering `wgpu`/`wgpu-hal`?

As for `0.52`: that really is quite a ways back and very unlikely to be version-range-compatible, especially for `Direct3D12`. It'll definitely work for our purposes, though. Let me know if you want me to downgrade in all PRs.

Comment (glandium) [reply Cargo.lock:1525]: `windows-sys` 0.52 is what's used by plenty of crates. `windows` 0.52 is only used by gpu-allocator. Note the versions are *NOT* correlated. 0.52 is the current version for `windows-sys` and `windows` 0.58 depends on it. Anyways, there is some manual work involved in upgrading `windows` in mozilla-central, but nothing prevents it from happening.

Comment (MarijnS95) [reply Cargo.lock:1525]: @glandium note that `windows` doesn't depend on `windows-sys`. It depends on `windows-targets`. Which `windows-sys` also depends on.

It is (unfortunately) not a classical `-sys` setup that helps deduplicating low-level definitions by splitting them out of the `windows` crate into `windows-sys` (i.e. having (vtable) function definitions there would be immensely helpful).

Comment (glandium) [reply Cargo.lock:1525]: Sure, but that doesn't change the version requirements for mozilla-central.

File: wgpu-hal/Cargo.toml
Comment (cwfitzgerald) [reply wgpu-hal/Cargo.toml]: For context on the workspace table, we introduced it, but there was a bug in cargo-vendor that caused Firefox to not be able to injest it, so we reverted core, hal, and types (the only crates ff uses) back to standard dependencies. The bug has been long fixed, I just don't think anyone has put in the work to verify it's fine and move back to a workspace dep.

**Comment by cwfitzgerald** [comment]
I suspect this windows version dance is going to be a recurring theme in our future. That is one upside off using a years-abandoned crate, nothing ever changes 😛

Anywho I think we (the maintainers) should figure out what the viable paths forward are that satisfy mozilla-central and wgpu's external users and then clearly communicate that, as I don't want Marijn to do duplicate work and I don't want this to explode mozilla-central.

- Ideally we can figure out a way to land 0.57/58. It's not improbable that nothing else in the stack breaks and mozilla-central can just update, though I know there will be politics around that
- if not, dropping back and supporting as wide of a range as we can, to minimize duplicate dependencies. We should earmark Marjin's improvements, so when et do update, we can re-implement them. 
- loom is indeed irrelevant
- if there is no range beyond 0.52

File: wgpu-hal/Cargo.toml
Comment (waywardmonkeys) [reply wgpu-hal/Cargo.toml]: I've opened #6020 to start cleaning up some of the workspace dependencies (if that's helpful).

Comment (MarijnS95) [reply wgpu-hal/Cargo.toml]: Thanks @waywardmonkeys, looking forward to see that PR fix at least the dependencies used here, so that this PR can simply be based on top. Looks like you have a lot of dependencies to go, though.

File: Cargo.lock
Comment (MarijnS95) [reply Cargo.lock:1525]: > 0.52 is only used by gpu-allocator.

`gpu-allocator` is a crate that I maintain and recently released a new version for with support all the way up to `windows 0.58`.

The `mozilla-central` repo indeed [uses `windows 0.52`](https://searchfox.org/mozilla-central/rev/de46cd99a7d1634058b511a3f546a970763e5048/Cargo.lock#6915-6916) [for `gpu-allocator`](https://searchfox.org/mozilla-central/rev/de46cd99a7d1634058b511a3f546a970763e5048/Cargo.lock#2562) but it's [only used by `wgpu-hal`](https://searchfox.org/mozilla-central/rev/de46cd99a7d1634058b511a3f546a970763e5048/Cargo.lock#6819) whose version I am bumping **in this PR**, so this should be a **non-issue**.

It seems you [have a semver-compatible patch for `windows-targets`](https://searchfox.org/mozilla-central/rev/de46cd99a7d1634058b511a3f546a970763e5048/Cargo.toml#110-112) anyway.

**Review by glandium — APPROVED** [review]
No problem for me to upgrade to windows 0.58.

File: wgpu-hal/Cargo.toml
Comment (MarijnS95) [reply wgpu-hal/Cargo.toml]: > > Note that default-features only turns off the std feature: I'm not exactly sure what that implies but we may get worse error formatting - at least it was like that years ago.
> 
> Good to know, maybe worth a comment on it in case you can confirm? Generally good practice for a library like wgpu not to drag in default features when they're not needed even if it's something like improved error handling, after all a user of wgpu can still do that just fine. But the other way round a user of wgpu wouldn't be able to opt-out of those default features at all

Upon closer inspection, [the feature is broken](https://github.com/microsoft/windows-rs/issues/3163) ([fix here](https://github.com/microsoft/windows-rs/pull/3164)).

- `windows-core`: Only uses it to turn on `Event`, which is a wrapper for delegates. The feature is **not forwarded to windows-result` nor `windows-strings`** (which is a bug I'd say);
- `windows`: only forwards it to `windows-core`.

The transitive crates:

- `windows-result`: implements `std::error::Error` for `Error`, and conversions to/from `std::io::Error` (which is very similar on Windows);
- `windows-strings`: various conversion for `HSTRING` to and from `OsStr(ing)` and `Path`.

In the end it's nothing that seems to be hiding nicer internals such as error formatting from us. Everything that it exposes is explicit public API (and if we end up relying on it, turning off `std` will cause compiler errors).

File: Cargo.lock
Comment (cwfitzgerald) [reply Cargo.lock:1525]: Heh, nice!

**Comment by MarijnS95** [comment]
@cwfitzgerald since we've made the resolution to just go ahead and use `windows 0.58` directly (it seems), can we get reviews finished? Landing this PR first will make it much easier and cleaner to rebase #5956 on top.

**Review by cwfitzgerald — APPROVED** [review]
Great stuff! Got the all clear to merge from the moz team.

**Comment by cwfitzgerald** [comment]
Ready to land, just needs merge conflicts resolved

**Pull request merged**
